utf8/reader_test.go

```diff
@@ -0,0 +1,288 @@
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch")
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
+			}
+		}
+	}
+
+	// TODO(i4k): implement stdiotest.TestReader for utf8.Reader
+}
+
+func TestUTF8ReaderMultipleSizes(t *testing.T) {
+	input := iotest.NewRepeatReader(bytes.NewBuffer([]byte(expectedString)), 100)
+	inputBytes, err := io.ReadAll(input)
+	assert.NoError(t, err, "failed to repeat input")
+
+	for i := 0; i < 10; i++ {
+		buf1 := bytes.NewBuffer(inputBytes)
+		buf2 := bytes.NewBuffer(inputBytes)
+		repeater1 := iotest.NewRepeatReader(buf1, i)
+		repeater2 := iotest.NewRepeatReader(buf2, i)
+
+		runes, err := runes.ReadAll(utf8.NewReaderReader(repeater1))
+		assert.NoError(t, err, "reading runes")
+
+		expectedBytes, err := io.ReadAll(repeater2)
+		assert.NoError(t, err, "reading expected runes")
+
+		expected := []rune(string(expectedBytes))
+
+		assert.EqualInts(t, len(expected), len(runes), "length mismatch")
+
+		for j := 0; j < len(expected); j++ {
+			if expected[j] != runes[j] {
+				t.Errorf("rune %d are not equal: %c != %c",
+					j, expected[j], runes[j])
+			}
+		}
+	}
+}
+
+func TestUTF8ReaderHalfRead(t *testing.T) {
+	input := iotest.NewRepeatReader(bytes.NewBuffer([]byte(expectedString)), 100)
+	inputBytes, err := io.ReadAll(input)
+	assert.NoError(t, err, "failed to repeat input")
+
+	for i := 0; i < 10; i++ {
+		buf1 := bytes.NewBuffer(inputBytes)
+		buf2 := bytes.NewBuffer(inputBytes)
+		repeater1 := stdiotest.HalfReader(iotest.NewRepeatReader(buf1, i))
+		repeater2 := iotest.NewRepeatReader(buf2, i)
+
+		runes, err := runes.ReadAll(utf8.NewReaderReader(repeater1))
+		assert.NoError(t, err, "reading runes")
+
+		expectedBytes, err := io.ReadAll(repeater2)
+		assert.NoError(t, err, "reading expected runes")
+
+		expected := []rune(string(expectedBytes))
+
+		assert.EqualInts(t, len(expected), len(runes), "length mismatch")
+
+		for j := 0; j < len(expected); j++ {
+			if expected[j] != runes[j] {
+				t.Errorf("rune %d are not equal: %c != %c",
+					j, expected[j], runes[j])
+			}
+		}
+	}
+}
+
+func TestUTF8ReaderFromFileMultipleBufferSizes(t *testing.T) {
+	expectedRunes := []rune(expectedString)
+	expectedLen := len(expectedRunes)
+
+	temp, err := ioutil.TempFile("", "spells-utf8")
+	assert.NoError(t, err, "creating temp file")
+
+	defer os.Remove(temp.Name())
+
+	n, err := temp.WriteString(expectedString)
+	assert.NoError(t, err, "writing utf8 file")
+	assert.EqualInts(t, len(expectedString), n, "written len mismatch")
+
+	for i := 0; i < expectedLen; i++ {
+		t.Run(fmt.Sprintf("using buffer size: %d", i), func(t *testing.T) {
+			off, err := temp.Seek(0, 0)
```

Comment:
Hmm what about we:

1 - Close temp after writing the contents, outside the for loop/subtest
2 - On each subtest, open the file again (read only), no need for seeking
3 - Since each subtest opens the file, we could use t.Parallel() on the subtests

Thinking more on the whole tests taking too much time issue, not sure if this test is one of the worse offenders, just an idea. Besides performance, I would also find it more isolated/easier to understand (clear stage of writing/closing + read only on each subtest, less chance of one test messing up another one, etc).